### PR TITLE
Chroma key example

### LINKED ISSUE #80: Example of chrome keying a live video.
- Live video input
- Color picker to determine which color to key out.
- Threshold parameter

### LINKED ISSUE #219: Add support for VideoFrames

### PR DESCRIPTION
**Pull request by reczkok**
Closes #80 
Closes #219 

- [x] Fix editor types

### CODE REVIEW

File: apps/typegpu-docs/src/content/examples/image-processing/chrome-keying.ts [hunk 1]
```diff
@@ -0,0 +1,229 @@
+/*
+{
+  "title": "Chrome Keying",
```
Comment (mhawryluk) [inline apps/typegpu-docs/src/content/examples/image-processing/chrome-keying.ts]:
```suggestion
  "title": "Chroma Keying",
```

\+ change the name of the file

File: apps/typegpu-docs/package.json [hunk 2]
```diff
@@ -18,6 +18,7 @@
     "@monaco-editor/react": "^4.6.0",
     "@radix-ui/react-select": "^2.1.1",
     "@radix-ui/react-slider": "^1.2.0",
+    "@types/dom-mediacapture-transform": "^0.1.9",
```
Comment (mhawryluk) [inline apps/typegpu-docs/package.json:21]: not visible in the examples code editor. probably needs to be passed as extraLib to tsDefaults in `CodeEditor.tsx` just like webgpu types

File: apps/typegpu-docs/src/content/examples/image-processing/chrome-keying.ts [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+const computeProgram = runtime.makeComputePipeline({
+  code: wgsl`
+    let coords = vec2u(${builtin.globalInvocationId}.xy);
+    if (coords.x >= ${width} || coords.y >= ${height}) {
+      return;
+    }
+    let xyAsUv = vec2f(coords) / vec2f(${width}, ${height});
+    var col = textureSampleBaseClampToEdge(${externalTexture}, ${sampler}, xyAsUv);
+    let toKey = ${colorBuffer.asUniform()};
```
Comment (mhawryluk) [inline apps/typegpu-docs/src/content/examples/image-processing/chrome-keying.ts]: i don't think this variable is necessary, and this name is a little confusing. also, please merge the main branch and use asUniform standalone function here

File: apps/typegpu-docs/src/content/examples/image-processing/chrome-keying.ts [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+      const uv = array(
+        vec2(1.0, 0.0),
+        vec2(1.0, 1.0),
+        vec2(0.0, 1.0),
+        vec2(1.0, 0.0),
+        vec2(0.0, 1.0),
+        vec2(0.0, 0.0),
+      );
+
+      let Position = vec4(pos[${builtin.vertexIndex}], 0.0, 1.0);
```
Comment (mhawryluk) [inline apps/typegpu-docs/src/content/examples/image-processing/chrome-keying.ts]: `position`

File: apps/typegpu-docs/src/content/examples/image-processing/chrome-keying.ts [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+});
+
+video.addEventListener('click', (event) => {
+  const { offsetX: x, offsetY: y } = event;
+  const canvas = document.createElement('canvas');
+  const context2d = canvas.getContext('2d');
+  if (!context2d) throw new Error('2d context not supported');
+
+  canvas.width = width;
+  canvas.height = height;
```
Comment (mhawryluk) [inline apps/typegpu-docs/src/content/examples/image-processing/chrome-keying.ts]: creating this canvas on every click seems excessive. I suggest moving it above to the global scope. though this way we keep the reference to this helper canvas forever and ever and it cannot be garbage collected, but I still think it's better. what do you think?

File: packages/typegpu/src/wgslBindGroupResolver.ts [hunk 6]
```diff
@@ -225,14 +225,18 @@ export class BindGroupResolver {
...
-        (ext) => ext.descriptor.source instanceof HTMLVideoElement,
-      )
-    ) {
-      this.invlidateBindGroup();
-    }
-    // future invalidation logic will go here
+    for (const texture of this.externalTextures) {
+      // check if texture is dirty (changed source) -> if so, invalidate bind group and mark clean
+      if (this.runtime.isDirty(texture)) {
+        this.invlidateBindGroup();
```
Comment (mhawryluk) [inline packages/typegpu/src/wgslBindGroupResolver.ts]: => invalidateBindGroup

File: packages/typegpu/src/wgslTexture.ts [hunk 7]
```diff
@@ -60,7 +64,9 @@ export interface WgslTextureView<
 }
 
 export interface WgslTextureExternal extends WgslRenderResource {
-  readonly descriptor: GPUExternalTextureDescriptor;
+  $name(label: string): WgslTextureExternal;
```
Comment (mhawryluk) [inline packages/typegpu/src/wgslTexture.ts]: merge main and use `WgslNamable` interface

File: apps/typegpu-docs/src/content/examples/image-processing/chrome-keying.ts (hunk 5 above)
Comment (reczkok) [reply apps/typegpu-docs/src/content/examples/image-processing/chrome-keying.ts]: I don't think this is much of a problem either way. The clicks will not be happening often enough to cause any real performance issues. The approach with one canvas is probably a little cleaner - I will change it.

File: apps/typegpu-docs/src/components/CodeEditor.tsx [hunk 8]
```diff
@@ -23,15 +23,39 @@ const typegpuExtraLibs = pipe(
...
   tsDefaults.addExtraLib(webgpuTypes);
   for (const lib of typegpuExtraLibs) {
     tsDefaults.addExtraLib(lib.content, lib.filename);
   }
+  for (const lib of mediacaptureExtraLibs) {
+    tsDefaults.addExtraLib(lib.content, lib.filename);
+  }
   tsDefaults.addExtraLib(toolkitTypes, 'example-toolkit.d.ts');
   tsDefaults.addExtraLib(typedBinary, 'typed-binary.d.ts');
+  tsDefaults.addExtraLib(typedBinary, 'typed-binary.d.ts');
```
Comment (mhawryluk) [inline apps/typegpu-docs/src/components/CodeEditor.tsx]: i think once is enough. ✂️

File: apps/typegpu-docs/src/content/examples/image-processing/chrome-keying.ts [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+  primitive: {
+    topology: 'triangle-list',
+  },
+});
+
+video.addEventListener('click', (event) => {
+  const { offsetX: x, offsetY: y } = event;
+
+  samplingCanvas.width = width;
+  samplingCanvas.height = height;
```
Comment (mhawryluk) [inline apps/typegpu-docs/src/content/examples/image-processing/chrome-keying.ts]: probably can be moved out of this function and therefore assigned just once

**Pull request merged**